### ISSUE TITLE
[Question] How to properly train-infer transformer on padded batches of sequences?

### ISSUE THREAD

**Issue by Penchekrak** [espnet/espnet#4940]
I wonder how to tackle the following problem, which arise when I train on batches of padded sequences:
- [relative positional encoding](https://github.com/espnet/espnet/blob/master/espnet/nets/pytorch_backend/transformer/embedding.py#L260) is dependent on sequence length and for padded batch it is only valid for longest sequence in batch
- `rel_shift` [operation](https://github.com/espnet/espnet/blob/master/espnet/nets/pytorch_backend/transformer/attention.py#L237) also seem to work correctly only for longest sequence in intermediate attention matrix.
Suppose now, that training on padded batches, without any modifications to the above code, went fine: when I perform inference of such model on individual sequences, those positional encodings differ from what model have seen during training. Can this discrepancy be somehow mitigated?

**Comment by sw005320** [comment]
@pengchengguo, can you answer this question?
I'm also curious.

**Comment by Penchekrak** [comment]
Let me elaborate slightly on this issue.
Consider padded batch with following exemplar mask:
![image](https://user-images.githubusercontent.com/1496107/219317699-2e3da85e-d044-434f-9ac8-04b53120ea31.png)
masked relative positional encoding for batched computation looks like (only zeroth channel is displayed here and below)
![image](https://user-images.githubusercontent.com/1496107/219317935-5f44772a-bb7d-4d8f-abf0-cb490d9c1249.png)
whereas positional encoding for individual element-wise computation after appropriate padding and stacking together looks like 
![image](https://user-images.githubusercontent.com/1496107/219318383-da07effb-1562-4f58-bf65-3a03e30b7db1.png)
For `matrix_bd` inside attention layer it is more or less the same story: 
`rel_shift` output for batched matrices looks like (again, only zeroth channel, non-vanishing entries of tensors are displayed)
![image](https://user-images.githubusercontent.com/1496107/219318779-d351f74f-2063-4081-b443-55af93d2f383.png)
(which looks closer to what is described in original paper)
but for stacked matrices for element-wise computation it is
![image](https://user-images.githubusercontent.com/1496107/219319015-e246d50e-5790-4173-80cb-f2989db14c60.png)


I believe this can affect model performance considerably, so I'm eager to find, how to make individual and batched computations more similar

**Comment by pengchengguo** [comment]
Hi @Penchekrak,

I am trying to follow your comments and sorry I do not understand how to get the third figure.
I know the second figure is obtained by applying a mask to the positional embedding (E_{emb} * Mask). Its horizontal axis refers to the time dimension, while the vertical axis is the batch index.
But for "positional encoding for individual element-wise computation after appropriate padding and stacking together", what does this step means?

**Comment by Penchekrak** [comment]
> I am trying to follow your comments and sorry I do not understand how to get the third figure. 
> "positional encoding for individual element-wise computation after appropriate padding and stacking together", what does this step means?

Sure, thank you for a prompt response!
So the way I have obtained third picture was like so -- I applied `RelPositionalEncoding` to each element of the batch separately, only accounting for their original length, just like one would do at inference time where there are no batches.
Approximate code snippet is:
```
batch_new = torch.zeros(batch.shape[0], batch.shape[1], self.attention_dim).to(batch.device)
for i in range(len(masks)):
    seq_len = masks[i].sum().int().item()
    out = pos_enc_layer(batch[[i], :seq_len])
    batch_new[I, :seq_len] = out.squeeze()
```

**Comment by pengchengguo** [comment]
I see, please give me some time to think about it. I will reply as soon as possible.

**Comment by Penchekrak** [comment]
@pengchengguo, please, take a look at this possible fix for discovered issue https://github.com/espnet/espnet/pull/4966

**Comment by pengchengguo** [comment]
Hi @Penchekrak,
Thanks so much for your contribution, I will check it.

**Comment by pengchengguo** [comment]
> > I am trying to follow your comments and sorry I do not understand how to get the third figure.
> > "positional encoding for individual element-wise computation after appropriate padding and stacking together", what does this step means?
> 
> Sure, thank you for a prompt response! The way I have obtained third picture was as follows -- I applied `RelPositionalEncoding` to each element of the batch separately, only accounting for their original length, just like one would do at inference time where there are no batches. Approximate code snippet is:
> 
> ```
>  # second image
> pos_enc = pos_enc_layer(batch).expand(batch.shape[0], -1, -1) * pos_enc_mask
> 
> # third image
> pos_enc = torch.zeros(batch.shape[0], batch.shape[1], self.attention_dim).to(batch.device)
> for i in range(len(masks)):
>     seq_len = masks[i].sum().int().item()
>     out = pos_enc_layer(batch[[i], :seq_len])
>     pos_enc[I, :seq_len] = out.squeeze()
> ```
> 
> Stacking them back together is just a way to demonstrate, how, at inference time, those entries will have different pos_enc applied to them, compared to training batched regime

Hi @Penchekrak,

I try to compute positional embedding using the following code:

```python
import torch

from espnet.nets.pytorch_backend.e2e_asr import pad_list
from espnet.nets.pytorch_backend.nets_utils import make_pad_mask
from espnet.nets.pytorch_backend.transformer.embedding import RelPositionalEncoding


# given the following setup
attn_dim = 256
xs_len = [20, 20, 19, 18, 18, 17, 17, 15, 15, 14, 13, 13, 12, 12, 11, 11, 10, 10]

# 1. simulate data
xs_lst = []
for l in xs_len:
    xs_lst.append(torch.rand(l, attn_dim))
xs_pad = pad_list(xs_lst, 0)  # (B, T, D)
xs_len = torch.tensor(xs_len)
masks = ~make_pad_mask(xs_len)[:, None, :]

# 2. init pos_enc_layer
pos_enc_layer = RelPositionalEncoding(attn_dim, dropout_rate=0.0)

# 3. compute positional embedding by batch (like your figure 2)
# here, pos_emb_batch.shape = (1, 2*T-1, D), 2T-1 means we use a combination of [pe_positive, pe_negative]
# suppose `i` means to the position of query vecotr and `j` means the position of key vector. 
# we use position relative positions when i >j (history information) and negative relative positions when i < j (future information).
xs_pad, pos_emb_batch = pos_enc_layer(xs_pad)

# 4. compute positional embedding one-by-one (like your figure 3)
pos_emb_indiv_lst = []
for x in xs_lst:
    # use x.unsqueeze(0) because pos_enc_layer takes a tensor of (B, T, *)
    tmp_x, tmp_pos_emb = pos_enc_layer(x.unsqueeze(0))
    pos_emb_indiv_lst.append(tmp_pos_emb)
```

For your figure 3, we can't directly draw it because the obtained pos_enc refers to different relative positions, for example
```
pos_emb_indiv_lst[1] means i-j \in (-20, 20)
pos_emb_indiv_lst[2] means i-j \in (-19, 19)
pos_emb_indiv_lst[3] means i-j \in (-18, 18)
```

After printing the `pos_emb_indiv_lst`, we can find the red part are same:
<img width="736" alt="image" src="https://user-images.githubusercontent.com/23135280/226108176-af21a031-9fb2-40c1-824a-1b4b604b9dbe.png">

I am not sure whether I explain this clearly,  please feel free to tell me if I have some misunderstands.
